File: backend/routes/download.js
const express = require('express');
const router = express.Router();
const jwt = require('jsonwebtoken');
const pool = require('../db/pool');

// Middleware to verify JWT token
const verifyToken = (req, res, next) => {
  const authHeader = req.headers.authorization;
  if (!authHeader || !authHeader.startsWith('Bearer ')) {
    return res.status(401).json({ error: 'No token provided' });
  }

  const token = authHeader.split(' ')[1];
  try {
    const decoded = jwt.verify(token, process.env.JWT_SECRET);
    req.userId = decoded.userId;
    next();
  } catch (err) {
    return res.status(401).json({ error: 'Invalid token' });
  }
};

// Middleware to verify active license
const verifyLicense = async (req, res, next) => {
  try {
    const result = await pool.query(
      `SELECT * FROM licenses
       WHERE user_id = $1
       AND active = true
       AND (expires_at IS NULL OR expires_at > NOW())`,
      [req.userId]
    );

    if (result.rows.length === 0) {
      return res.status(403).json({ error: 'Active subscription required' });
    }

    req.license = result.rows[0];
    next();
  } catch (err) {
    console.error('License verification error:', err);
    return res.status(500).json({ error: 'Server error' });
  }
};

// Download URLs - replace with your actual download URLs
const DOWNLOAD_URLS = {
  mac: process.env.DOWNLOAD_URL_MAC || '/downloads/PokerSharkScope-mac.zip',
  windows: process.env.DOWNLOAD_URL_WINDOWS || '/downloads/PokerSharkScope-windows.zip'
};

// GET /api/download/mac
router.get('/download/mac', verifyToken, verifyLicense, (req, res) => {
  // Log the download
  console.log(`User ${req.userId} downloading Mac version`);

  // Redirect to download URL
  res.redirect(DOWNLOAD_URLS.mac);
});

// GET /api/download/windows
router.get('/download/windows', verifyToken, verifyLicense, (req, res) => {
  // Log the download
  console.log(`User ${req.userId} downloading Windows version`);

  // Redirect to download URL
  res.redirect(DOWNLOAD_URLS.windows);
});

module.exports = router;

File: backend/app/api/user/download.py
"""Download endpoints for the app."""

from datetime import datetime, timezone
from pathlib import Path
from typing import Optional
from fastapi import APIRouter, HTTPException, Depends, status, Query
from fastapi.responses import FileResponse, RedirectResponse
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select
from sqlalchemy.orm import selectinload
from jose import jwt, JWTError

from app.services.settings_service import SettingsService
from app.database import get_session
from app.models import User, License, LicenseStatus
from app.config import get_settings

settings = get_settings()

router = APIRouter(tags=["Downloads"])

# Local downloads folder
DOWNLOADS_DIR = Path(__file__).parent.parent.parent.parent / "downloads"


async def get_user_from_token(token: str, session: AsyncSession) -> User:
    """Get user from JWT token."""
    try:
        payload = jwt.decode(token, settings.secret_key, algorithms=["HS256"])
        user_id = payload.get("sub")
        if not user_id:
            raise HTTPException(status_code=401, detail="Invalid token")

        result = await session.execute(
            select(User).options(selectinload(User.licenses)).where(User.id == user_id)
        )
        user = result.scalar_one_or_none()
        if not user:
            raise HTTPException(status_code=401, detail="User not found")
        return user
    except JWTError:
        raise HTTPException(status_code=401, detail="Invalid token")


@router.get("/download/{platform}")
async def download_app(
    platform: str,
    token: Optional[str] = Query(None),
    session: AsyncSession = Depends(get_session),
):
    """Download the app for the specified platform. Requires login and active license."""
    if platform not in ("mac", "windows"):
        raise HTTPException(status_code=400, detail="Invalid platform. Use 'mac' or 'windows'")

    # Get user from token
    if not token:
        raise HTTPException(status_code=401, detail="Authentication required")

    current_user = await get_user_from_token(token, session)

    # Check if user has any licenses
    if not current_user.licenses:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="You need an active license to download. Please purchase a subscription."
        )

    # Find an active license
    active_license = None
    now = datetime.now(timezone.utc)
    for lic in current_user.licenses:
        if lic.status == LicenseStatus.ACTIVE:
            if not lic.expires_at:
                active_license = lic
                break
            # Make expires_at timezone aware if it's naive
            exp = lic.expires_at
            if exp.tzinfo is None:
                exp = exp.replace(tzinfo=timezone.utc)
            if exp >= now:
                active_license = lic
                break

    if not active_license:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Your license is not active or has expired. Please renew your subscription."
        )

    # Check for local file first
    if platform == "mac":
        local_file = DOWNLOADS_DIR / "PokerSharkScope-Mac.zip"
        filename = "PokerSharkScope-Mac.zip"
    else:
        local_file = DOWNLOADS_DIR / "PokerSharkScope-Windows.zip"
        filename = "PokerSharkScope-Windows.zip"

    # If local file exists, serve it directly
    if local_file.exists():
        return FileResponse(
            path=str(local_file),
            filename=filename,
            media_type="application/zip"
        )

    # Otherwise, check for external URL in settings
    settings_service = SettingsService(session)
    downloads = await settings_service.get_setting_value("downloads", {})

    if platform == "mac":
        url = downloads.get("mac_url")
    else:
        url = downloads.get("windows_url")

    if url:
        return RedirectResponse(url=url)

    raise HTTPException(
        status_code=404,
        detail=f"Download for {platform} is not available. Please contact support."
    )
